lib/firestore/delete.js

```diff
@@ -280,18 +273,120 @@ FirestoreDelete.progressBar = new ProgressBar("Deleted :current docs (:rate docs
+  var maxQueueSize = deleteBatchSize * maxPendingDeletes * 2;
+
+  // All temporary variables for the deletion queue.
+  var queue = [];
+  var numPendingDeletes = 0;
+  var pagesRemaining = true;
+  var pageIncoming = false;
+  var lastDocName;
+
+  var failures = [];
+  var retried = {};
+
+  var queueLoop = function() {
+    if (queue.length == 0 && numPendingDeletes == 0 && !pagesRemaining) {
+      return true;
+    }
+
+    if (failures.length > 0) {
+      logger.debug("Found " + failures.length + " failed deletes, failing.");
+      return true;
     }
 
-    return self._deleteDocuments(docs).then(function(numDeleted) {
-      // Tick the progress bar
-      FirestoreDelete.progressBar.tick(numDeleted);
+    if (queue.length <= maxQueueSize && pagesRemaining && !pageIncoming) {
+      pageIncoming = true;
 
-      // Recurse to delete another batch
-      return self._recursiveBatchDelete();
-    });
+      self
+        ._getDescendantBatch(self.allDescendants, readBatchSize, lastDocName)
+        .then(function(docs) {
+          pageIncoming = false;
+
+          if (docs.length == 0) {
+            pagesRemaining = false;
+            return;
+          }
+
+          docs.forEach(function(doc) {
+            queue.push(doc);
+          });
+
+          lastDocName = docs[docs.length - 1].name;
+        })
+        .catch(function(e) {
+          logger.debug("Failed to fetch page after " + lastDocName, e);
+          pageIncoming = false;
+        });
+    }
+
+    if (numPendingDeletes > maxPendingDeletes) {
+      return false;
+    }
+
+    if (queue.length == 0) {
+      return false;
+    }
+
+    var toDelete = [];
+    var numToDelete = Math.min(deleteBatchSize, queue.length);
+
+    for (var i = 0; i < numToDelete; i++) {
+      toDelete.push(queue.shift());
+    }
+
+    numPendingDeletes++;
+    firestore
+      .deleteDocuments(self.project, toDelete)
+      .then(function(numDeleted) {
+        FirestoreDelete.progressBar.tick(numDeleted);
+        numPendingDeletes--;
+      })
+      .catch(function(e) {
+        // For server errors, retry if the document has not yet been retried.
+        if (e.status >= 500 && e.status < 600) {
+          logger.debug("Server error deleting doc batch", e);
+
+          // Retry each doc up to one time
+          toDelete.forEach(function(doc) {
+            if (retried[doc.name]) {
+              logger.debug("Failed to delete doc " + doc.name + " multiple times.");
+              failures.push(doc.name);
+            } else {
+              retried[doc.name] = true;
+              queue.push(doc);
+            }
+          });
+        } else {
+          logger.debug("Fatal error deleting docs ", e);
+          toDelete.forEach(function(doc) {
+            failures.push(doc.name);
+          });
+        }
+
+        numPendingDeletes--;
+      });
+
+    return false;
+  };
+
+  return new Promise(function(resolve, reject) {
+    var intervalId = setInterval(function() {
```

Comment:
Can we not make this a normal loop (maybe by using setImmediate)?

Reply:
Can you explain how?  My naive attempt was like this:

```js
    var done = false;
    while (!done) {
      setImmediate(function() {
        done = queueLoop();
      });
    }
```

But that does nothing.

Reply:
It's not immediately obvious to me why you need to call setImmediate(). Do you have Promise-chains that otherwise do not get resolved?

If you need the setImmediate, you could try something like this:

```
let continue = function() {
    if (queueLoop()) {
      return Promise.resolve();
    } else {
      return new Promise(resolve => continue().then(resolve()));
   }
}

return continue();
```

I am not sold that this is cleaner or better. It is probably mostly just different.

Reply:
I don't really know enough JS to use the right words, but I need to make sure that the promises within `queueLoop` actually get executed, rather than being blocked forever by a tight loop.

So something like this:
```
    while(!done) {
      done = queueLoop();
    }
```

Never does any work.  I assume I could be doing something with `process.nextTick()` but I'm not sure what.

Reply:
Anyway I've tried about 20 different varieties and determined:

  * This one works
  * The loop type is _not_ the limiting factor in performance.  The limiting factor is (surprisingly) read speed, at least on smallish data sets without index hotspots.

Reply:
Sounds good to me. Thanks for trying :)